File: 2024/elixir/day3.exs
defmodule Main do

  def read_input() do
    File.read!("../inputs/day3.txt")
  end


  def helper(str ,i) do

    str = String.slice(str, i+4..-1)
      |> String.split(")", trim: true)
      |> Enum.at(0)
      |> String.split(",", trim: true)

    if Enum.count(str) == 2 do

      case Integer.parse(Enum.at(str, 0)) do
        {num, ""} -> num
        _ -> 0
      end * case Integer.parse(Enum.at(str, 1)) do
        {num, ""} -> num
        _ -> 0
      end

    else
      0
    end

  end

  
  def part1() do

    input = read_input()

    ans = Enum.sum(for i <- 0..String.length(input) do

      if String.at(input, i) == "m" && String.at(input, i+1) == "u" && String.at(input, i+2) == "l" && String.at(input, i+3) == "(" do
        helper(input, i)
      else
        0
      end

    end)

    IO.puts("The answer for part 1 is: #{ans}")
    
  end


  def part2() do

    input = read_input()

    ans = Enum.reduce(0..String.length(input), {0,true}, fn i, {acc,boo} ->

      boo = if String.at(input, i) == "d" && String.at(input, i+1) == "o" && String.at(input, i+2) == "(" && String.at(input, i+3) == ")" do
        true
      else
        boo
      end

      boo = if String.at(input, i) == "d" && String.at(input, i+1) == "o" && String.at(input, i+2) == "n" && String.at(input, i+3) == "'"
            && String.at(input, i+4) == "t" && String.at(input, i+5) == "(" && String.at(input, i+6) == ")" do
        false
      else
        boo
      end

      if boo && String.at(input, i) == "m" && String.at(input, i+1) == "u" && String.at(input, i+2) == "l" && String.at(input, i+3) == "(" do
        {acc + helper(input, i) , boo} 
      else
        { acc + 0 , boo }
      end

    end)

    IO.puts("The answer for part 2 is: #{elem(ans, 0)}")
    
  end

end

Main.part1()
Main.part2()

File: 2024/rust/src/bin/day3.rs
use std::fs;
use unicode_segmentation::UnicodeSegmentation;


fn main() ->  Result<(), Box<dyn std::error::Error>> {

    let input = fs::read_to_string("../inputs/day3.txt")?;

    let mut content: String = String::new();

    for i in input.lines(){
        content.push_str(i);
    }

    println!("The answer for part 1 is: {}", part1(&content));
    println!("The answer for part 2 is: {}", part2(&content));

    Ok(())
}   

fn part1 (content: &String) -> i64 {
    let content = content.graphemes(true).collect::<Vec<&str>>();

    let mut ans:i64 = 0;

    for i in 0..content.len() {
        if content[i] == "m" && content[i+1] == "u" && content[i+2] == "l" && content[i+3] == "(" {
            ans+=helper(i, &content);
        }
    }

    ans
}

fn part2 (content: &String) -> i64 {
    let content = content.graphemes(true).collect::<Vec<&str>>();

    let mut ans:i64 = 0;

    let mut doo:bool = true;

    for i in 0..content.len() {

        if content[i] == "d" && content[i+1] == "o" && content[i+2] == "(" && content[i+3] == ")" {
            doo = true;
        }

        if content[i] == "d" && content[i+1] == "o" && content[i+2] == "n" && content[i+3] == "'" && content[i+4] == "t" && content[i+5] == "(" && content[i+6] == ")" {
            doo = false;
        }

        if doo && content[i] == "m" && content[i+1] == "u" && content[i+2] == "l" && content[i+3] == "(" {
            ans+=helper(i, &content);
        }
    }

    ans
}

fn helper (mut i: usize, content: &Vec<&str>) -> i64 {

    let mut num1: String = String::new();
    let mut num2: String = String::new();

    for j in i+4..content.len() {
        if content[j] == "," {
            i = j;
            break;
        }
        if content[j].parse::<i64>().is_ok() {
            num1.push_str(content[j]);
        }
        else{
            return 0;
        }
    }

    for j in i+1..content.len() {
        if content[j] == ")" {
            break;
        }
        if content[j].parse::<i64>().is_ok() {
            num2.push_str(content[j]);
        }
        else{
            return 0;
        }
    }

    let num1 = num1.parse::<i64>().unwrap();
    let num2 = num2.parse::<i64>().unwrap();

    num1*num2
}
